docs pass & omnigraph overview (#2142)
namehash/ensnode: docs/ensnode.io/src/content/docs/docs/integrate/omnigraph/index.mdx

---
title: ENS Omnigraph API

> description: A protocol-native, unified GraphQL API over indexed ENS data — polymorphic across ENSv1 and ENSv2.

Reviewer: ```suggestion
description: The world's first and only API providing unified access to the full state of ENSv1 and ENSv2.
```

---

import { LinkCard, CardGrid } from "@astrojs/starlight/components";

> The **ENS Omnigraph API** is the GraphQL API over the indexed Unigraph data model. It's a single, polymorphic schema over **both ENSv1 and ENSv2** — write your query once and get correct, typed results regardless of which protocol version a given Domain lives in.

Reviewer: ```suggestion
The **ENS Omnigraph API** is the world's first and only API providing unified access to the full state of ENSv1 and ENSv2. It's a single, polymorphic GraphQL API over **both ENSv1 and ENSv2** — write your query once and get correct, typed results regardless of which protocol version a given Domain lives in.
```

![ENS Omnigraph diagram](/ens-omnigraph-diagram.png)

The Omnigraph is delivered by [ENSApi](/docs/services/ensapi) on top of the indexed Unigraph in [ENSDb](/docs/services/ensdb). It follows the [Relay specification](https://relay.dev/graphql/connections.htm), abstracts away the most common ENS-protocol footguns, and exposes enough of the underlying protocol for builders who need to go deep.

## One unified API over ENSv1 + ENSv2

> Both ENSv1 and ENSv2 Domains are indexed concurrently and exposed through the same schema. When ENSv2 launches in **Summer 2026**, the two protocol versions coexist — and the Omnigraph keeps your app working against both, at the same time, with no code changes.

Reviewer: ```suggestion
When ENSv2 launches in **Summer 2026**, the two protocol versions coexist — and the Omnigraph keeps your app working against both, at the same time, with no code changes. Both ENSv1 and ENSv2 Domains are indexed concurrently and exposed through a unified schema.
```

> Both ENSv1 and ENSv2 Domains are indexed concurrently and exposed through the same schema. When ENSv2 launches in **Summer 2026**, the two protocol versions coexist — and the Omnigraph keeps your app working against both, at the same time, with no code changes.

Reviewer: @shrugs Can you please make a special effort to explicitly highlight a critical detail about ENSv2 that I think many will miss.

Many will interpret the current narrative here and elsewhere in the docs that the Omnigraph API will give all ENSv1 domains until ENSv2 launches and then give all ENSv2 domains after ENSv2 launches. But the reality is much more complex, right? ENSv1 domains continue to be ENSv1 domains after ENSv2 launches. They optionally become ENSv2 domains only when their owner decides to upgrade them to ENSv2 after ENSv2 launches. Therefore, both ENSv1 and ENSv2 domains coexist concurrently.

Could you please take the lead to communicate ideas like this explicitly in multiple places? All the places where someone may benefit from understanding this critical detail about ENSv2. We shouldn't assume readers will read our docs in a perfectly ordered way from page 1 -> N. We should repeat certain key ideas like this over and over again where relevant.

Author: Done — the ENSv1+ENSv2 coexistence framing is now reinforced in README, integrate/index, ensv2-readiness, why-ensnode, and the Omnigraph overview (incl. the L16 + L18 suggestions you proposed).

> Addressing a Domain by name (`domain(by: { name: "vitalik.eth" })`) prefers the ENSv2 entity when one exists, and falls back to the ENSv1 entity otherwise. Set-returning queries (`domains(where: { name: ... })`) return both protocol versions, unless explicitly filtered with `version: ENSv1` or `version: ENSv2`.

Reviewer: ```suggestion
Addressing a Domain by name (`domain(by: { name: "vitalik.eth" })`) returns an ENSv2 Domain if the domain has been optionally upgraded by its owner to ENSv2, and returns an ENSv1 Domain otherwise. Set-returning queries (`domains(where: { name: ... })`) return both the ENSv1 and ENSv2 onchain representations of the same domain, unless explicitly filtered with `version: ENSv1` or `version: ENSv2`.
```

## What is a Namegraph?

> A **Namegraph** is the native data model of ENSv2: it represents names not as a flat mapping of namehashes (as in ENSv1), but as a graph of `Registry → Domain → Registry → Domain → …`. This graph may be cyclic, and within the ENSv2 protocol an unbounded number of _disjoint_ (not connected) Namegraphs may/will exist. Within the unified ENS protocol (v1+v2) there will exist many Namegraphs, at the very least those headed by the ENSv1 Root Registry, Basenames, Lineanames, 3DNS, and the ENSv2 Root Registry.

Reviewer: ```suggestion
A **Namegraph** is the native onchain data model of ENSv2: it represents names not as a flat mapping of namehashes in a Nametable (as in ENSv1), but as a graph of `Registry → Domain → Registry → Domain → …`. This graph may be cyclic, and within the ENSv2 protocol an unbounded number of _disjoint_ (not connected) Namegraphs will exist. Within the unified ENS protocol (v1+v2) there will exist many Namegraphs, at the very least those headed by the ENSv1 Root Registry, Basenames, Lineanames, 3DNS, and the ENSv2 Root Registry.
```

> A **Namegraph** is the native data model of ENSv2: it represents names not as a flat mapping of namehashes (as in ENSv1), but as a graph of `Registry → Domain → Registry → Domain → …`. This graph may be cyclic, and within the ENSv2 protocol an unbounded number of _disjoint_ (not connected) Namegraphs may/will exist. Within the unified ENS protocol (v1+v2) there will exist many Namegraphs, at the very least those headed by the ENSv1 Root Registry, Basenames, Lineanames, 3DNS, and the ENSv2 Root Registry.

Reviewer: Can you please double check this?

> Within the unified ENS protocol (v1+v2) there will exist many Namegraphs, at the very least those headed by the ENSv1 Root Registry, Basenames, Lineanames, 3DNS, and the ENSv2 Root Registry.

My mental model is that:
- There's many disjoint Namegraphs, but only 1 canonical Namegraph.
- The canonical Namegraph is separate from ENSv1 Nametables, and therefore cannot be relied upon as the sole source of truth to determine the canonical Nametree (based on ENS protocol resolution mechanics).

Appreciate your advice for how we communicate these ideas 👍 Please proceed as you think is best.

Author: Applied your L22 suggestion verbatim and the L26 Unigraph-vs-Namegraph terminology fix in the same pass. Resolving.

## What is the Unigraph?

> The **Unigraph** is the entire collection of these disjoint namegraphs, indexed and stitched together using ENS Forward Resolution semantics, into a single navigable Namegraph. The result is a single, uniform data model — every ENSv1 query is the same shape as every ENSv2 query. Navigating the Namegraph from `"eth"` down to `"vitalik.eth"` and beyond looks identical regardless of whether the underlying entities are ENSv1 or ENSv2.

Reviewer: ```suggestion
The **Unigraph** is the entire collection of these disjoint ENSv2 Namegraphs and multiple ENSv1 Nametables, combined together into a single unified data model using ENS Resolution semantics. Navigating the Unigraph from `"eth"` down to `"vitalik.eth"` and beyond looks identical regardless of whether the underlying entities are ENSv1 or ENSv2.
```

@shrugs Hey please note the key terminology feedback included in this comment, including how we should refer to the indexed data model as the Unigraph not the Namegraph. Appreciate there's so many ...graphs.

> The [`unigraph` plugin](/docs/services/ensindexer) in ENSIndexer is what builds this unified model. It's also where multichain coverage lives: Basenames (`.base.eth`), Lineanames (`.linea.eth`), and 3DNS names (`.box`) are all materialized into the same Namegraph as mainnet `.eth`, so a single query covers all indexable names in one shot.

Reviewer: ```suggestion
The [`unigraph` plugin](/docs/services/ensindexer) in ENSIndexer is what builds this unified model. It's also where multichain coverage lives: Basenames (`.base.eth`), Lineanames (`.linea.eth`), and 3DNS names (`.box`) are all materialized into the same Unigraph as mainnet `.eth`, so a single query covers all indexable names in one shot.
```

## Canonicality

> Given that a Domain entity (say, the `sub` in `sub.example.eth`) can be reached by infinitely many aliases (for example, `sub.other.eth`), it becomes important to determine a _canonical_ reference to the Domain — this is the **Canonical Name**. Canonicality is also connected to nameability; if a Domain exists on-chain but isn't eventually connected to the Root Registry somehow, **it doesn't have a name!**

Reviewer: ```suggestion
Given that a Domain entity (say, the `sub` in `sub.example.eth`) can be reached by infinitely many aliases (for example, `sub.other.eth`), it becomes important to determine a _canonical_ reference to the Domain — this is the **Canonical Name**. Canonicality is also connected to nameability; if an ENSv2 Domain exists on-chain but isn't eventually connected to the Root Registry somehow, **it doesn't have a name!**
```

Within the Omnigraph API the complexity of this is reduced, and all Canonical Domains (those with names) are queryable, searchable, and addressable by said name. Domains that are not canonical are _still_ referenceable by `id` (eg. `domain(by: { id: DomainId! })`).

> Canonical Domains have a `Domain.canonical` field hosting the canonicality-derived fields such as `name`, `node`, `depth` (i.e. 2 for `vitalik.eth`), and `path` (i.e. `[Domain("eth"), Domain("vitalik.eth")]`).

Reviewer: ```suggestion
Canonical Domains have a `Domain.canonical` field hosting the canonicality-derived fields such as `name`, `node`, `depth` (i.e. 2 for `vitalik.eth`), and `path` from the ENS root (i.e. `[Domain("eth"), Domain("vitalik.eth")]`).
```

## Stable IDs vs. Namegraph addressing

> Every entity in the Omnigraph has an `id` — a nominally-typed, stable reference to a specific on-chain entity (`DomainId`, `RegistryId`, `RegistrationId`, etc.). When you already know the entity you mean, address it by `id`: `domain(by: { id: "..." })` always returns the same on-chain Domain.

Reviewer: ```suggestion
Every entity in the Omnigraph has an `id` — a nominally-typed, stable reference to a specific on-chain entity (`DomainId`, `RegistryId`, `RegistrationId`, etc.). When you already know the entity you mean, address it by `id`: `domain(by: { id: "..." })` always returns the same on-chain Domain, even if its canonical name changes across time.
```

> Addressing a Domain by **name** is a different operation. It's **Namegraph forward traversal**: `domain(by: { name: "vitalik.eth" })` walks from the ENSv2 Root Registry → `"eth"` in that Registry → the Registry that `"eth"` points at (the EthRegistry) → `"vitalik"` in that Registry. The Domain returned is whichever on-chain entity that path currently resolves to.

Reviewer: ```suggestion
Addressing a Domain by **name** is a different operation. It's **Unigraph forward traversal**: `domain(by: { name: "vitalik.eth" })` walks from the ENSv2 Root Registry → `"eth"` in that Registry → the Registry that `"eth"` points at (the EthRegistry) → `"vitalik"` in that Registry. The Domain returned is whichever on-chain entity (if any) that path currently resolves to.
```

> Addressing a Domain by **name** is a different operation. It's **Namegraph forward traversal**: `domain(by: { name: "vitalik.eth" })` walks from the ENSv2 Root Registry → `"eth"` in that Registry → the Registry that `"eth"` points at (the EthRegistry) → `"vitalik"` in that Registry. The Domain returned is whichever on-chain entity that path currently resolves to.

Author: Outdated against the current text. The Omnigraph overview is scoped to the unigraph plugin, where the ENSv2 Root Registry is canonical; ENSv1-only namespace behavior is covered separately. Resolving.

These two views are not interchangeable:

- The `id` you receive from a name lookup _is_ the stable reference to the on-chain entity that the Namegraph currently resolves `"vitalik.eth"` to.
- But `"vitalik.eth"` is not a stable reference to that entity. The Namegraph can be re-parented or re-aliased — and tomorrow, `"vitalik.eth"` may resolve to an entirely different on-chain Domain.

**Rule of thumb:** address by `name` when you're answering "which Domain would records come from if resolved right now"; address by `id` when you're answering "what's the latest state of this specific on-chain entity?".

## Polymorphism via GraphQL interfaces

`Domain`, `Registry`, and `Registration` are GraphQL **interfaces**, with concrete types implementing each:

- `Domain` → `ENSv1Domain`, `ENSv2Domain`
- `Registry` → `ENSv1Registry`, `ENSv1VirtualRegistry`, `ENSv2Registry`
- `Registration` → `BaseRegistrarRegistration`, `NameWrapperRegistration`, `ThreeDNSRegistration`, `ENSv2RegistryRegistration`, `ENSv2RegistryReservation`

> Shared fields are available unconditionally on the interface. Protocol- or implementation-specific fields are reached via typed inline fragments — `... on ENSv1Domain { rootRegistryOwner }`, `... on ENSv2Domain { tokenId }`, `... on BaseRegistrarRegistration { wrapped { fuses } }`, `... on NameWrapperRegistration { fuses }`. The result is a single query that compiles, type-checks, and returns the right fields for whichever concrete type each row turns out to be.

Reviewer: ```suggestion
Shared fields are available unconditionally on the interface. Protocol- or implementation-specific fields are reached via typed inline fragments — `... on ENSv1Domain { rootRegistryOwner }`, `... on ENSv2Domain { tokenId }`, `... on BaseRegistrarRegistration { wrapped { fuses } }`, `... on NameWrapperRegistration { fuses }`. The result is a single query that compiles, type-checks, and returns the right fields for whichever concrete type each record turns out to be.
```

## Interpreted Names everywhere

> Every name and label crossing the Omnigraph surface is an **Interpreted Name** (or Interpreted Label). Each label in an Interpreted Name is either a normalized literal label or an Encoded LabelHash (`[abc123…]`) when the literal isn't known. This eliminates the most common ENS UI footguns — unnormalized labels, unhealed hashes, and rendering surprises — at the schema layer, making UI rendering trivial. See [terminology](/docs/reference/terminology#interpreted-label) for the full definition.

Reviewer: ```suggestion
Every name and label crossing the Omnigraph surface is an **Interpreted Name** (or Interpreted Label). Each label in an Interpreted Name is either a normalized literal label or an Encoded LabelHash (`[abc123…]`) when the literal isn't known or is unnormalized. This eliminates one of the most common ENS UI footguns — unnormalized labels, unhealed hashes, and rendering surprises — at the schema layer, making UI rendering trivial. See [terminology](/docs/reference/terminology#interpreted-label) for the full definition.
```

## Relay-spec connections

> Every list field in the schema is a [Relay-spec Connection](https://relay.dev/graphql/connections.htm) with `edges`, `pageInfo`, and `totalCount`. Cursor-based pagination is idiomatic in urql, Apollo, Relay, and most modern GraphQL clients — infinite scroll and stable pagination work out of the box, with no per-endpoint plumbing.

Author: Applied in a0780627d — narrowed to `Collection and paginated relationship fields…` and called out `DomainCanonical.path` / `Event.topics` as plain-array exceptions.

> Every list field in the schema is a [Relay-spec Connection](https://relay.dev/graphql/connections.htm) with `edges`, `pageInfo`, and `totalCount`. Cursor-based pagination is idiomatic in urql, Apollo, Relay, and most modern GraphQL clients — infinite scroll and stable pagination work out of the box, with no per-endpoint plumbing.

Author: Duplicate of the other VADE thread on this line — fixed in a0780627d.

## A complete audit log of ENS Events

The Omnigraph indexes every onchain Event relevant to ENS and exposes it from the entities each Event relates to:

- `Domain.events` — every Event for a specific Domain
- `Resolver.events` — every Event emitted by a specific Resolver
- `Account.events` — every Event for which an Account is the HCA-aware `sender`
- `Permissions.events`, `PermissionsUser.events` — Permission grant and revocation history

Each `Event` carries chain, block, transaction, and log metadata, plus an HCA-aware `sender` field distinct from the raw `tx.from` for HCA-mediated transactions.

> ## First-class Permissions

Reviewer: Seems it would be nice to include a reference to how to query the fuses of wrapped ENSv1 domains. Could you add it?

Author: The L59 Polymorphism update now references both `BaseRegistrarRegistration { wrapped { fuses } }` and `NameWrapperRegistration { fuses }` inline. Resolving — happy to add a worked example query in a follow-up if that would help.
